### PR TITLE
redistribute bedrock and stripe usd

### PR DESCRIPTION
**Pull request by GNaD13**
## Description

Closes: #997 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by faddat** [comment]
Differences between this and #1029 ?

**Comment by nghuyenthevinh2000** [comment]
> Differences between this and #1029 ?

it is different ser. Just wrong title for this pr. Dang work is on redistribute bedrock and stripe USD. While, Nhan is on setting up code structure.

File: app/upgrades/v4/upgrades.go [hunk 1]
```diff
@@ -1,27 +1,210 @@
...
+		}
+	}
+
+	// Send 1 ubedrock to each validator
+	vals := staking.GetAllValidators(ctx)
+	for _, val := range vals {
+		_, err = staking.Delegate(
+			ctx,
+			sdk.MustAccAddressFromBech32(EngineHotWal),
+			math.OneInt(),
```
Comment (nghuyenthevinh2000) [inline app/upgrades/v4/upgrades.go:149]: it has to be 1000000 ubedrock = 1 bedrock

File: app/upgrades/v4/upgrades.go [hunk 2]
```diff
@@ -1,27 +1,210 @@
...
+		IncentivePools,
+		FoundationDiscretionary,
+		TkPreSale,
+		TkCFSale,
+		TkRegDSSale,
+		CompanyRevenue,
+		EngineHotWal,
+	}
+
+	TotalUbedrock = math.NewIntFromUint64(1_000_000_000)
```
Comment (nghuyenthevinh2000) [inline app/upgrades/v4/upgrades.go:49]: change ubedrock to bedrock (10^6)

**Review by nghuyenthevinh2000 — CHANGES_REQUESTED** [review]

File: app/upgrades/v4/upgrades.go (hunk 1 above)
Comment (GNaD13) [reply app/upgrades/v4/upgrades.go]: fixed

File: app/upgrades/v4/upgrades.go (hunk 2 above)
Comment (GNaD13) [reply app/upgrades/v4/upgrades.go]: fixed

File: app/upgrades/v4/upgrades.go [hunk 3]
```diff
@@ -1,27 +1,210 @@
...
+		TkPreSale,
+		TkCFSale,
+		TkRegDSSale,
+		CompanyRevenue,
+		EngineHotWal,
+	}
+
+	TotalUbedrock = math.NewIntFromUint64(1_000_000_000_000_000) // 1 bedrock = 1_000_000 ubedrock
+
+	UbedrockDistribute = map[string]math.Int{
```
Comment (mustafapylons) [inline app/upgrades/v4/upgrades.go:51]: All the bedrock numbers should be in constant variables.

File: app/upgrades/v4/upgrades.go [hunk 4]
```diff
@@ -1,27 +1,210 @@
...
+			// Burn denom token in module
+			err = bank.BurnCoins(ctx, types.PaymentsProcessorName, sdk.NewCoins(amount))
+			if err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// Mint ubedrock for 8 security account
```
Comment (mustafapylons) [inline app/upgrades/v4/upgrades.go:122]: //mint ubedrock for 8 security accounts.

File: app/upgrades/v4/upgrades_test.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/stretchr/testify/suite"
+)
+
+var (
+	stakingCoinDenom string    = "ubedrock"
+	stripeCoinDenom  string    = "ustripeusd"
+	defaultAcctFunds sdk.Coins = sdk.NewCoins(
+		sdk.NewCoin(stakingCoinDenom, sdk.NewInt(10_000_000)),
```
Comment (mustafapylons) [inline app/upgrades/v4/upgrades_test.go:19]: let's use constants for these as well.

**Review by mustafapylons — COMMENTED** [review]
Few small issues, otherwise code is good.

File: app/upgrades/v4/upgrades.go (hunk 4 above)
Comment (GNaD13) [reply app/upgrades/v4/upgrades.go:122]: sorry sir, what is this problem?

Comment (GNaD13) [reply app/upgrades/v4/upgrades.go:122]: i mean, what should i do to resolve this issue??

**Pull request merged**